### PR TITLE
Check read permissions

### LINKED ISSUE #56: Check field permissions for GET request
Currently the script doesn't check if the user is permitted to view the fields on a page (custom permission). Is something like this implemented or on the todo list already?

### PR DESCRIPTION
**Pull request by casdr**
Closes #56

### CODE REVIEW

**Comment by tisto** [comment]
Hi @casdr, thanks for your contribution! Before we can review your pull request you need to resolve your merge conflicts. In addition we would require some basic tests before we can merge. Let us know if you need any assistance.

File: src/plone/restapi/deserializer/atcontent.py [hunk 1]
```diff
@@ -24,6 +24,7 @@ def __init__(self, context, request):
         self.request = request
 
     def __call__(self, validate_all=False):
+
```
Comment (maartenkling) [inline src/plone/restapi/deserializer/atcontent.py]: remove

File: src/plone/restapi/serializer/__init__.py [hunk 2]
```diff
@@ -90,10 +96,34 @@ def SerializeToJson(context):
             }
             for member in context.results()
         ]
-
+
```
Comment (maartenkling) [inline src/plone/restapi/serializer/__init__.py]: remove

File: src/plone/restapi/services/content/get.py [hunk 3]
```diff
@@ -1,7 +1,8 @@
 # -*- coding: utf-8 -*-
 from plone.rest import Service
 from plone.restapi.interfaces import ISerializeToJson
-
+from AccessControl import Unauthorized
```
Comment (maartenkling) [inline src/plone/restapi/services/content/get.py]: check pyflakes

File: src/plone/restapi/serializer/__init__.py [hunk 4]
```diff
@@ -90,10 +96,34 @@ def SerializeToJson(context):
...
     for schema in iterSchemata(context):
         for field_name, field in getFields(schema).items():
-            value = lookup_field_serializer(context, field)()
-            result[json_compatible(field_name)] = value
+            read_permissions = mergedTaggedValueDict(
+                schema, READ_PERMISSIONS_KEY)
+
+            if not check_permission(context, read_permissions.get(field_name)):
+                continue
+            else:
```
Comment (maartenkling) [inline src/plone/restapi/serializer/__init__.py]: dont else

**Comment by tisto** [comment]
@casdr I just checked and it seems you haven't signed the Plone contributor agreement. This is necessary before we can merge your code. Please see https://plone.org/foundation/contributors-agreement for details.

**Comment by maartenkling** [comment]
@tisto i will take care of the agreement!

**Comment by tisto** [comment]
@maartenkling cool. So I guess @casdr is a fourdigits guy. :)

File: buildout.cfg [hunk 5]
```diff
@@ -1,2 +1,3 @@
 [buildout]
+allow-picked-versions = true
```
Comment (maartenkling) [inline buildout.cfg]: remove

File: docs/source/_json/collection.json [hunk 6]
```diff
@@ -14,12 +14,7 @@ content-type: application/json
     "test_user_1_", 
     "admin"
   ], 
-  "customViewFields": [
-    "Title", 
-    "Creator", 
-    "Type", 
-    "ModificationDate"
-  ], 
+  "customViewFields": [],
```
Comment (maartenkling) [inline docs/source/_json/collection.json]: Why is this changed?

File: docs/source/_json/event.json [hunk 7]
```diff
@@ -27,7 +27,7 @@ content-type: application/json
   "expires": null, 
   "language": "", 
   "location": null, 
-  "open_end": false, 
+  "open_end": null,
```
Comment (maartenkling) [inline docs/source/_json/event.json]: Why is this changed?

File: docs/source/_json/event.json [hunk 8]
```diff
@@ -42,5 +42,5 @@ content-type: application/json
   "text": null, 
   "timezone": null, 
   "title": "Event", 
-  "whole_day": false
+  "whole_day": null
```
Comment (maartenkling) [inline docs/source/_json/event.json]: Why is this changed?

File: pip-selfcheck.json [hunk 9]
```diff
@@ -0,0 +1 @@
+{"last_check":"2016-02-09T16:17:39Z","pypi_version":"8.0.2"}
```
Comment (maartenkling) [inline pip-selfcheck.json]: put in gitignore

File: src/plone/restapi/services/content/get.py [hunk 10]
```diff
@@ -2,7 +2,6 @@
 from plone.rest import Service
 from plone.restapi.interfaces import ISerializeToJson
 
-
```
Comment (maartenkling) [inline src/plone/restapi/services/content/get.py]: put me back

File: src/plone/restapi/tests/dxtypes.py [hunk 11]
```diff
@@ -4,7 +4,7 @@
 from datetime import timedelta
 from plone.app.textfield import RichText
 from plone.app.vocabularies.catalog import CatalogSource
-from plone.autoform.directives import write_permission
+from plone.autoform.directives import write_permission, read_permission
```
Comment (maartenkling) [inline src/plone/restapi/tests/dxtypes.py]: dont do import using ,

File: versions.cfg [hunk 12]
```diff
@@ -11,4 +11,8 @@ plone.recipe.codeanalysis = 2.0b1
 Sphinx = 1.3.1
 docutils = 0.12
 Pygments = 2.0
-six = 1.4
\ No newline at end of file
+six = 1.4
+check-manifest = 0.31
+mccabe = 0.4.0
+pyflakes = 1.0.0
+collective.recipe.sphinxbuilder = 0.8.2
```
Comment (maartenkling) [inline versions.cfg]: new pins?

File: src/plone/restapi/serializer/dxfields.py [hunk 13]
```diff
@@ -28,6 +28,8 @@ def __init__(self, field, context, request):
         self.context = context
         self.request = request
         self.field = field
+        self.permission_cache = {}
```
Comment (maartenkling) [inline src/plone/restapi/serializer/dxfields.py]: why?

File: src/plone/restapi/serializer/__init__.py [hunk 14]
```diff
@@ -93,7 +98,30 @@ def SerializeToJson(context):
...
+
+            if not check_permission(context, read_permissions.get(field_name)):
+                continue
             value = lookup_field_serializer(context, field)()
             result[json_compatible(field_name)] = value
 
     return result
+
+
+def check_permission(context, permission_name):
```
Comment (maartenkling) [inline src/plone/restapi/serializer/__init__.py]: Is this code dublicated from: https://github.com/plone/plone.restapi/blob/44ac5b4242349d99fe65ba6f89400cd955a3e8ce/src/plone/restapi/deserializer/dxcontent.py#L113 why not import it and reuse?

**Comment by maartenkling** [comment]
@tisto correct, contract is send

File: versions.cfg [hunk 15]
```diff
@@ -11,4 +11,3 @@ plone.recipe.codeanalysis = 2.0b1
 Sphinx = 1.3.1
 docutils = 0.12
 Pygments = 2.0
-six = 1.4
```
Comment (maartenkling) [inline versions.cfg]: why is this deleted?

**Comment by maartenkling** [comment]
@tisto @casdr has been accepted please review the pull if you like

**Comment by buchi** [comment]
@casdr thanks for your contribution! Can you please squash your commits?

**Comment by casdr** [comment]
@tisto @buchi I've squashed the commits

File: docs/source/_json/collection.json [hunk 16]
```diff
@@ -5,62 +5,62 @@ HTTP 200 OK
 content-type: application/json
 
 {
-  "@context": "http://www.w3.org/ns/hydra/context.jsonld",
```
Comment (buchi) [inline docs/source/_json/collection.json]: It's not obvious to me why this file has changes. The examples are generated when the tests are executed and have a trailing whitespace. Did you remove the trailing whitespaces manually?

File: docs/source/_json/event.json [hunk 17]
```diff
@@ -5,42 +5,42 @@ HTTP 200 OK
 content-type: application/json
 
 {
-  "@context": "http://www.w3.org/ns/hydra/context.jsonld",
```
Comment (buchi) [inline docs/source/_json/event.json]: Same here, this file should not change.

File: src/plone/restapi/serializer/dxfields.py [hunk 18]
```diff
@@ -11,7 +11,6 @@
 from zope.interface import Interface
 from zope.schema.interfaces import IField
 
-
```
Comment (buchi) [inline src/plone/restapi/serializer/dxfields.py]: Please do not change that file as the change is not related to the added functionality.

File: versions.cfg [hunk 19]
```diff
@@ -11,4 +11,4 @@ plone.recipe.codeanalysis = 2.0b1
...
\ No newline at end of file
+six = 1.4
```
Comment (buchi) [inline versions.cfg]: Please keep this file as is. Cleanup should be done in seperate commits, though it's not necessary here.

File: src/plone/restapi/serializer/__init__.py [hunk 20]
```diff
@@ -18,6 +18,12 @@
 from zope.component import queryMultiAdapter
 from zope.schema import getFields
 
+from plone.autoform.interfaces import READ_PERMISSIONS_KEY
```
Comment (buchi) [inline src/plone/restapi/serializer/__init__.py]: Please keep imports sorted (case sensitive). Grouping is not necessary.

**Comment by casdr** [comment]
I think the documentation (automatically generating by running test) is causing the branch conflict, aren't these files meant to be ignored?

**Comment by buchi** [comment]
The docs are generated automatically an depending on the Plone version they differ a little bit. In your commit simply do not change any of the files in `docs/source/_json`. Then there can't be a conflict caused by these files.

Your commit should only change the following files:
- `.gitignore`
- `src/plone/restapi/serializer/__init__.py`
- `src/plone/restapi/tests/dxtypes.py`
- `src/plone/restapi/tests/test_serializer.py`

**Comment by casdr** [comment]
Closed in favor of PR [64](https://github.com/plone/plone.restapi/pull/64l)
